src/Quantum/Quantum.jl

```diff
@@ -84,17 +96,14 @@ Base.hcat(tns::TensorNetwork...) = reduce(hcat, tns)
 function Base.adjoint(tn::TensorNetwork{A}) where {A<:Quantum}
     tn = deepcopy(tn)
 
-    # TODO refactor internals
-    for i in siteinds(tn)
-        plug = i.meta[:plug]
-        i.meta[:plug] = if plug == :input
-            :output
-        elseif plug == :output
-            :input
-        else
-            # TODO throw error?
-        end
-    end
+    tmp = Dict((site, if dir === :in
+        :out
+    elseif dir === :out
+        :in
+    else
+        dir
+    end) => index for ((site, dir), index) in tn[:plug])
+    merge!(tn[:plug], tmp)
```

Comment:
```suggestion
    keys_to_modify = []
    for ((site, state), value) in tn[:plug]
        if state == :in
            push!(keys_to_modify, ((site, :in), value, (site, :out)))
        elseif state == :out
            push!(keys_to_modify, ((site, :out), value, (site, :in)))
        end
    end

    for (key, value, new_key) in keys_to_modify
        tn[:plug][new_key] = value
        delete!(tn[:plug], key)
    end
```
Something like this will work.

Reply:
Why not just do
```suggestion
    tn[:plug] = Dict((site, if dir === :in
        :out
    elseif dir === :out
        :in
    else
        dir
    end) => index for ((site, dir), index) in tn[:plug])
```

Does this work?

Reply:
No... Because these structs are immutable and thus we can not reassign a field, we can just modify in-place these fields

Reply:
`TensorNetwork` is immutable so we can not do something like `tn.metadata = Dict(...)`. But `tn.metadata` is a `Dict` which is a mutable type, so things like `tn.metadata[:plug] = Dict(...)` are completely allowed.

The problem probably is that I haven't implemented `setindex(tn::TensorNetwork, key::Symbol, value)` but `tn.metadata[:plug] = ...` must work.